File: QytetetJava/src/modeloqytetet/Calle.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package modeloqytetet;


public class Calle extends Casilla{
    
    private int numHoteles = 0;
    private int numCasas = 0;
    private TituloPropiedad titulo;
    
    Calle (int numeroCasilla, int coste, int numHoteles, int numCasas, TituloPropiedad titulo) {
       super(numeroCasilla,coste);
       this.numHoteles = numHoteles;
       this.numCasas = numCasas;
       this.titulo = titulo;
       this.titulo.setCasilla(this);     
    }
    
     int getNumHoteles() {
       return numHoteles;
    }

    int getNumCasas() {
       return numCasas;
    }
    
    public TituloPropiedad getTituloPropiedad() {
       return titulo;
    }
    
   public boolean soyEdificable(){
       return !titulo.getHipotecada();
    }

    void setNumHoteles(int numHoteles) {
       this.numHoteles = numHoteles;
    }

    void setNumCasas(int numCasas) {
       this.numCasas = numCasas;
    }

    private void setTitulo(TituloPropiedad titulo) {
       this.titulo = titulo;
    }
    
    int costeAlquilerBase(){
        return titulo.getAlquilerBase();
    } 
    
     public TituloPropiedad asignarPropietario(Jugador propietario){

       titulo.setPropietario(propietario);
       return titulo;
    }
    
    int calcularValorHipoteca(){
        return titulo.getHipotecaBase() + (int)(numCasas * 0.5 * titulo.getAlquilerBase()+ this.numHoteles  * titulo.getAlquilerBase());
    }
    
    int cancelarHipoteca(){
       titulo.setHipotecada(false);
       int cantidadRecibida = this.calcularValorHipoteca();
       
       return cantidadRecibida;
    }
    
    int cobrarAlquiler(){
        int costeAlquilerBase;
        int costeAlquiler;
        
        costeAlquilerBase =  titulo.getAlquilerBase();
        costeAlquiler = costeAlquilerBase + (int) (numCasas * 0.5 + numHoteles * 2);
   
        titulo.cobrarAlquiler(costeAlquiler);    
        
        return costeAlquiler;
    }
    
    public int edificarCasa(){
        int costeEdificarCasa;
        setNumCasas(numCasas+1);
        costeEdificarCasa = titulo.getPrecioEdificar();
        
        return costeEdificarCasa;
    }
    
    int edificarHotel(){
        setNumHoteles(numHoteles+1);
        int costeEdificarHotel = titulo.getPrecioEdificar();

        return costeEdificarHotel;
    }
    
    boolean estaHipotecada(){
        return titulo.getHipotecada();
    }
    
    int getCosteHipoteca(){
       int cantidad = titulo.getHipotecaBase() + (int) (numCasas * 0.5 * titulo.getHipotecaBase()) + (numHoteles * titulo.getHipotecaBase());
       return cantidad;
    }
    
    int getPrecioEdificar(){
        return titulo.getPrecioEdificar();
    }
    
    int hipotecar(){
        titulo.setHipotecada(true);
        int cantidadRecibida = this.calcularValorHipoteca();
       
        return cantidadRecibida;
    }
    
    int precioTotalComprar(){
        return titulo.getAlquilerBase();
    }
    
    Boolean propietarioEncarcelado(){
        return titulo.propietarioEncacelado();
    }
    
    Boolean sePuedeEdificarCasa(int factorEspeculador){
        return getNumCasas()<4*factorEspeculador;

    }
    
    Boolean sePuedeEdificarHotel(int factorEspeculador){
        Boolean aux = this.getNumHoteles()<4*factorEspeculador; 
        if(aux){
            numCasas-=4;
        }
        return aux;

    }
    
     Boolean tengoPropietario(){
       return titulo.tengoPropietario();
    }

    int venderTitulo(){
        
        titulo.setPropietario(null);
        this.setNumHoteles(0);
        this.setNumCasas(0);
        int precioCompra = coste + (numCasas + numHoteles ) * titulo.getPrecioEdificar();
        int precioVenta =  (int)( precioCompra + titulo.getFactorRevaloracion()*precioCompra);
        
        return precioVenta;    
    }
    
    @Override
    public String toString(){
        return " numero: " + getNumeroCasilla()  + ",\n coste: " + getCoste() + titulo +",\n numCasas=" + numCasas + ",\n numHoteles=" + numHoteles + "\n";
    }
    
}

File: QytetetRuby/lib/calle.rb
#encoding: utf-8
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

 require_relative "tipo_casilla"

module ModeloQytetet

  class Calle < Casilla



    attr_accessor :num_hoteles, :num_casas, :titulo

      def initialize (nc,c,tit)
        super(nc,c,TipoCasilla::CALLE)
        @num_hoteles = 0
        @num_casas = 0
        @titulo = tit
      end

      def soy_edificable
         true 
      end

      def coste_alquiler_base
        @titulo.alquiler_base
      end

      def asignar_propietario(jugador)
        @titulo.set_propietario(jugador)
        titulo
      end

      def calcular_valor_hipoteca
        hipoteca_base=@titulo.hipoteca_base      
        (hipoteca_base + @num_casas * 0.5 * hipoteca_base + @num_hoteles * hipoteca_base).to_i
      end

      def cancelar_hipoteca
        @titulo.hipotecado=false
        calcular_valor_hipoteca      
      end

      def cobrar_alquiler
        coste_alquiler = coste_alquiler_base + (@num_casas * 0.5 + @num_hoteles * 2 ).to_i
        @titulo.cobrar_alquiler(coste_alquiler)
        coste_alquiler
      end

      def edificar_casa
        @num_casas+=1
        @titulo.precio_edificar 
      end

      def edificar_hotel
        @num_casas-=4
        @num_hoteles+=1
        @titulo.precio_edificar
      end

      def esta_hipotecada
        @titulo.hipotecado
      end

      def get_coste_hipoteca
        titulo.hipoteca_base + num_casas * 0.5 * titulo.hipoteca_base + num_hoteles * titulo.hipoteca_base
      end

      def get_precio_edificar
        titulo.precio_edificar      
      end 

      def hipotecar
        @titulo.hipotecado=true
        calcular_valor_hipoteca
      end    

      def precio_total_comprar
        titulo.alquiler_base
      end

      def propietario_encarcelado
        @titulo.propietario_encarcelado
      end

      def se_puede_edificar_casa(factor_especulador)
        @num_casas< (4 * factor_especulador)
      end

      def se_puede_edificar_hotel(factor_especulador)
        @num_hoteles<= (4 * factor_especulador) && @num_casas>=4
      end

      def tengo_propietario
        @titulo.tengo_propietario 
      end   

      def vender_titulo
        @titulo.propietario=nil
        @num_hoteles = 0
        @num_casas = 0

        precio_compra = coste + (num_casas + num_hoteles) * titulo.precio_edificar
        (precio_compra + titulo.factor_revalorizacion*precio_compra).to_i;
      end
      
      def to_s
           "  Numero de casilla: #{@numero_casilla}  coste: #{@coste}  Numero de hoteles: #{@num_hoteles}  Numero de casas: #{@num_casas}  Tipo: #{@tipo}  Titulo: #{@titulo}"
      end
  end
end
